edgedb/protocol/codecs/object.pyx

```diff
@@ -21,7 +21,74 @@
 cdef class ObjectCodec(BaseNamedRecordCodec):
 
     cdef encode(self, WriteBuffer buf, object obj):
-        raise NotImplementedError
+        cdef:
+            WriteBuffer elem_data
+            Py_ssize_t objlen
+            Py_ssize_t i
+            BaseCodec sub_codec
+
+        if not isinstance(obj, dict):
+            raise TypeError(
+                f'a dict is expected as a valid object argument, '
+                f'got {type(obj).__name__}')
+
+        self._check_encoder()
+
+        objlen = len(obj)
+        if objlen != len(self.fields_codecs):
+            print(type(obj), len(obj), obj)
+            raise self._make_missing_args_error_message(obj)
+
+        elem_data = WriteBuffer.new()
+        for i in range(objlen):
+            name = datatypes.record_desc_pointer_name(self.descriptor, i)
+            try:
+                arg = obj[name]
+            except KeyError:
+                raise self._make_missing_args_error_message(obj) from None
+
+            elem_data.write_int32(0)  # reserved bytes
+            if arg is None:
+                elem_data.write_int32(-1)
+            else:
+                sub_codec = <BaseCodec>(self.fields_codecs[i])
+                try:
+                    sub_codec.encode(elem_data, arg)
+                except (TypeError, ValueError) as e:
+                    value_repr = repr(arg)
+                    if len(value_repr) > 40:
+                        value_repr = value_repr[:40] + '...'
+                    raise errors.InvalidArgumentError(
+                        'invalid input for query argument'
+                        f' ${name}: {value_repr} ({e})') from e
+
+        buf.write_int32(4 + elem_data.len())  # buffer length
+        buf.write_int32(<int32_t><uint32_t>objlen)
+        buf.write_buffer(elem_data)
+
+    def _make_missing_args_error_message(self, args):
+        required_args = set()
+
+        for i in range(len(self.fields_codecs)):
+            name = datatypes.record_desc_pointer_name(self.descriptor, i)
+            required_args.add(name)
+
+        passed_args = set(args.keys())
+        missed_args = required_args - passed_args
+        extra_args = passed_args - required_args
+
+        error_message = f'expected {required_args} keyword arguments'
```

Comment:
```suggestion
        error_message = f'expected {required_args} arguments'
```

This isn't specific to keyword arguments, so mentioning them is confusing.  I don't think we need to be specific here, as we don't allow mixing argument styles, so simply saying "arguments" is enough.

Reply:
changed